### PR TITLE
Added support for code analysis of Maven projects with multiple modules.

### PR DESCRIPTION
**Pull request by HSAR**
This commit also removes the code cross-reference feature, which did not work without the relevant files anyway.

### CODE REVIEW

File: Tasks/Maven/analysisresult.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
...
 export class AnalysisResult {
     toolName: string;
-    filesWithViolations: number;
-    totalViolations: number;
-    xmlFilePath: string;
-    htmlFilePath: string;
+    filesWithViolations:number = 0;
+    totalViolations:number = 0;
+    xmlFilePath:string = null;
+    htmlFilePath:string = null;
```
Comment (HSAR) [inline Tasks/Maven/analysisresult.ts]: Default values allow the code to be less verbose.

File: Tasks/Maven/maventask.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ import Q = require('q');
 // Lowercased names are to lessen the likelihood of xplat issues
 import pmd = require('./pmdformaven');
 import ar = require('./analysisresult');
+import ma = require('./moduleanalysis');
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: See moduleanalysis.ts below.

File: Tasks/Maven/maventask.ts [hunk 3]
```diff
@@ -282,9 +306,15 @@ function applyPmdGoals(mvnRun: trm.ToolRunner):void {
     pmd.applyPmdArgs(mvnb);
 }
 
-// Returns true if any one code analysis tool was enabled.
-function isCodeAnalysisUploadRequired():boolean {
-    return tl.getInput('pmdAnalysisEnabled', true) == 'true';
+// Returns the names of any enabled code analysis tools, or empty array if none.
+function getEnabledCodeAnalysisTools():string[] {
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: This function was changed to provide an iterable list of enabled code analysis tools, as well as fulfilling its original function.

File: Tasks/Maven/maventask.ts [hunk 4]
```diff
@@ -282,9 +306,15 @@ function applyPmdGoals(mvnRun: trm.ToolRunner):void {
...
+    var result:string[] = [];
+
+    if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
+        result.push(pmd.toolName);
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: All tool names should be provided by the relevant external module. This is to reduce the fragility of the code.

File: Tasks/Maven/maventask.ts [hunk 5]
```diff
@@ -303,70 +333,108 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
...
-    tl.mkdirP(toolStagingDir);
+// Returns the full path of the staging directory for a given tool.
+function getStagingDirectory(toolName:string):string {
+    return path.join(stagingDir, toolName.toLowerCase());
+}
+
+// Copy output files to a staging directory and upload them as build artifacts.
+// Each module uploads its own build artifact.
+function uploadBuildArtifacts(toolName:string, moduleAnalysis:ma.ModuleAnalysis):void {
+    var localStagingDir:string = path.join(getStagingDirectory(toolName), moduleAnalysis.moduleName);
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: getStagingDirectory() was extracted into its own method to enhance the readability of this line, which is already quite complex.

File: Tasks/Maven/maventask.ts [hunk 6]
```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
...
+    if (enabledCodeAnalysisTools.length > 0) {
+        // Discover maven modules
+        var modules:ma.ModuleAnalysis[] = findModules(sourcesDir);
+
+        // Special case: if the root turns up as a module, the automatic name won't do
+        modules.forEach((module) => {
+            if (module.rootFolder == sourcesDir) {
+                module.moduleName = 'root'
+            }
+        });
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: The special case code should only be needed for single-module Maven projects. 

Multiple-module maven projects will have a pom.xml in the root, but should not have any compiled code associated with the master POM, and thus should have no /target/ folder.

Said special case code will protect even in the case of multiple-module maven projects with a module at the root.

File: Tasks/Maven/maventask.ts [hunk 7]
```diff
@@ -377,16 +447,23 @@ function uploadCodeAnalysisResults():void {
     }
 }
 
-// Take a result object and create one line of the build summary from it.
-function buildSummaryLineFromResult(analysisResult:ar.AnalysisResult):string {
+// For a given code analysis tool, create a one-line summary from multiple AnalysisResult objects.
+function createSummaryLine(toolName:string, analysisResults:ar.AnalysisResult[]):string {
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Renamed from "build" due to heavily overloaded meanings.

File: Tasks/Maven/moduleanalysis.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import ar = require('./analysisresult');
+// Data class for supporting Maven projects with more than one module
+export class ModuleAnalysis {
+    moduleName: string;
+    rootFolder: string;
+    analysisResults:any = {}; // To be used as a dictionary
+}
```
Comment (HSAR) [inline Tasks/Maven/moduleanalysis.ts]: New data class representing a Maven module. It has fields for the module name, the location of the root folder, and a dictionary of {string, AnalysisResult} used to associated toolName(s) with their analysisResult(s).

File: Tasks/Maven/pmdformaven.ts [hunk 9]
```diff
@@ -10,25 +10,27 @@ import trm = require('vsts-task-lib/toolrunner');
 
 import ar = require('./analysisresult');
 
+export const toolName:string = 'PMD';
+
 // Adds PMD goals, if selected by the user
 export function applyPmdArgs(mvnRun: trm.ToolRunner):void {
-    mvnRun.arg(['jxr:jxr', 'pmd:pmd']); // JXR source code cross-referencing is a required pre-goal for PMD
```
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Discovered that JXR source code cross-referencing is _not_ a required pre-goal, and will result in broken links unless the cross-reference files are uploaded as well.

File: Tasks/Maven/pmdformaven.ts [hunk 10]
```diff
@@ -10,25 +10,27 @@ import trm = require('vsts-task-lib/toolrunner');
...
+    mvnRun.arg(['pmd:pmd', '-DlinkXRef=false']); // Turn off cross-referencing to reduce Maven error output
 }
 
 // Extract analysis results from PMD output file.
-// Takes the current working directory and returns an AnalysisResult data class.
+// Takes the working directory (should contain pom.xml and target/) and returns an AnalysisResult data class.
 // Task fails if the HTML or XML outputs were not found.
 // @param sourcesDirectory - The absolute location of the root source directory.
-export function processPmdOutput(sourcesDirectory:string) : ar.AnalysisResult {
+export function processPmdOutput(rootDir:string) : ar.AnalysisResult {
```
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: No longer assumes we are working directly in the agent sources directory, but we still expect a /target/ dir.

File: Tasks/Maven/pmdformaven.ts [hunk 11]
```diff
@@ -37,43 +39,52 @@ export function processPmdOutput(sourcesDirectory:string) : ar.AnalysisResult {
 
 // Verifies the existence of the HTML output file.
 // Modifies the relevant field within the returned object accordingly.
-// Task fails if the file cannot be found.
 function processPmdHtml(analysisResult:ar.AnalysisResult, path:string):ar.AnalysisResult {
-    // Task fails if the file cannot be found.
     if (!tl.exist(path)) {
-        throw new Error("Could not find PMD HTML output. Check that PMD ran successfully and that it is writing to the default location.");
+        tl.debug('PMD HTML not found at ' + path);
```
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: No longer throws an error. See comment above about coding defensively in case module detection throws a false positive.

File: Tasks/Maven/pmdformaven.ts [hunk 12]
```diff
@@ -37,43 +39,52 @@ export function processPmdOutput(sourcesDirectory:string) : ar.AnalysisResult {
...
     var pmdXmlFileContents = fs.readFileSync(path, 'utf-8');
     xml2js.parseString(pmdXmlFileContents, function (err, data) {
-        if (data) {
-            analysisResult.xmlFilePath = path;
+        if (!data) { // Not an XML file, throw error
+            throw new Error("Failed to parse XML output from PMD.");
+        }
+        if (!data.pmd) { // Not a PMD XML, ignore
+            return analysisResult;
+        }
```
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Added additional checks to eliminate undefined/null ref errors and reorganised to throw/return early, avoiding excessive nested ifs.

File: Tasks/Maven/pmdformaven.ts [hunk 13]
```diff
@@ -37,43 +39,52 @@ export function processPmdOutput(sourcesDirectory:string) : ar.AnalysisResult {
...
-        throw new Error("Could not find PMD XML output. Check that PMD ran successfully and that it is writing to the default location.");
+        tl.debug('PMD XML not found at ' + path);
     }
 
     var pmdXmlFileContents = fs.readFileSync(path, 'utf-8');
     xml2js.parseString(pmdXmlFileContents, function (err, data) {
-        if (data) {
-            analysisResult.xmlFilePath = path;
+        if (!data) { // Not an XML file, throw error
+            throw new Error("Failed to parse XML output from PMD.");
```
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: I'm not sure it's necessary to throw here. This was never hit in development or testing.

File: Tests/L0/Maven/_suite.ts [hunk 14]
```diff
@@ -29,6 +29,48 @@ function setupDefaultMavenTaskRunner():tr.TaskRunner {
...
+
+    filesInFolder.forEach(function (fileInFolder) {
+        result.push(path.join(folder, fileInFolder));
+        if (fs.statSync(path.join(folder, fileInFolder)).isDirectory()) {
+            result = result.concat(listFolderContents(path.join(folder, fileInFolder)));
+        }
+    });
+
+    return result;
+}
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Convenience function used for setting up mock responses later.

File: Tests/L0/Maven/_suite.ts [hunk 15]
```diff
@@ -29,6 +29,48 @@ function setupDefaultMavenTaskRunner():tr.TaskRunner {
...
+
+    paths.forEach((path) => {
+        responseObject.exist[path] = true;
+        responseObject.checkPath[path] = true;
+        responseObject.rmRF[path] = rmRFSuccessObj;
+        responseObject.mkdirP[path] = true;
+    });
+
+    return responseObject;
+}
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Having to do this setup gets very bulky and obscures the point of tests, so has been moved here.

File: Tests/L0/Maven/_suite.ts [hunk 16]
```diff
@@ -44,58 +86,108 @@ describe('Maven Suite', function() {
     it('Maven / PMD: Executes PMD goals if PMD is enabled', (done) => {
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: This test is now the single-module test.

File: Tests/L0/Maven/_suite.ts [hunk 17]
```diff
@@ -44,58 +86,108 @@ describe('Maven Suite', function() {
...
+            })
+            .fail((err) => {
+                console.log(taskRunner.stdout);
+                console.log(taskRunner.stderr);
+                console.log(err);
+                done(err);
+            });
+    });
+
+    it('Maven / PMD: Detects and uploads results when multiple modules are present', (done) => {
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: New multi-module test.

File: Tests/L0/Maven/_suite.ts [hunk 18]
```diff
@@ -209,12 +307,15 @@ describe('Maven Suite', function() {
...
+                assert(taskRunner.ran('/usr/local/bin/mvn -f pom.xml package pmd:pmd -DlinkXRef=false'),
                     'should have run maven with the correct arguments');
 
                 done();
             })
             .fail((err) => {
+                console.log(taskRunner.stdout);
+                console.log(taskRunner.stderr);
+                console.log(err);
                 done(err);
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: General test changes - only log stdout and stderr to the console if the task fails, to avoid cluttering up other people's test runs.

File: Tests/L0/Maven/data/multimodule/util/target/pmd.xml [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+            Useless parentheses.
+        </violation>
+    </file>
+    <file name="C:\Agent\_work\3\s\src\main\java\test\test2.java">
+        <violation beginline="452" endline="452" begincolumn="22" endcolumn="36" rule="UselessParentheses" ruleset="Unnecessary" package="gamesincommon" class="Gui" method="addName" externalInfoUrl="https://pmd.github.io/pmd-5.3.5/pmd-java/rules/java/unnecessary.html#UselessParentheses" priority="4">
+            Useless parentheses.
+        </violation>
+        <violation beginline="452" endline="452" begincolumn="43" endcolumn="78" rule="UselessParentheses" ruleset="Unnecessary" package="gamesincommon" class="Gui" method="addName" externalInfoUrl="https://pmd.github.io/pmd-5.3.5/pmd-java/rules/java/unnecessary.html#UselessParentheses" priority="4">
+            Useless parentheses.
+        </violation>
```
Comment (HSAR) [inline Tests/L0/Maven/data/multimodule/util/target/pmd.xml]: All files below here (any pmd.xml and pom.xml files) can be ignored, as they are blank-ish test data. Their contents cannot be checked and so are not relevant.

File: Tasks/Maven/maventask.ts [hunk 20]
```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
...
+                    tl.exit(1);
                 }
-            } catch(e) {
-                tl.error(e.message);
-                tl.exit(1);
             }
-        }
 
-        // Process analysis results - upload files and generate build summary lines
+        });
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Each module has its own, independent /target/ folder, so each moduleAnalysis object is processed independently. The resulting AnalysisResult objects are stored in the analysisResults dictionary field of the moduleAnalysis object, associated by the toolname.

File: Tasks/Maven/maventask.ts [hunk 21]
```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
...
+                    if (moduleAnalysisResult) {
+                        toolAnalysisResults.push(moduleAnalysisResult);
+                    }
+
+                    uploadBuildArtifacts(toolName, module);
+                });
+
+                // After looping through all modules, summarise tool output results
+                var summaryLine:string = createSummaryLine(toolName, toolAnalysisResults);
+                buildSummaryLines.push(summaryLine);
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: While the gathering of files is more easily done per-module, the summary of results is done per-tool. Thus the second loop iterates over the enabled code analysis tools.

An inner for-each iterates through the module list. If the given module contained an AnalysisResult object for the current tool, we store the result and continue on to upload an artifact on a per-module, per-tool basis. 

After this, the summary line for the tool is created by iterating through all the AnalysisResult instances for the current tool and summing the total violations and files with violations.

**Comment by HSAR** [comment]
Follow up to an offline discussion - the implementation of the tasklib.rmRF() and mkdirP() functions are synchronous.

File: Tasks/Maven/maventask.ts [hunk 22]
```diff
@@ -272,6 +273,29 @@ function processMavenOutput(data) {
     }
 }
 
+// Identifies all maven modules below the root by the presence of a pom.xml file and a /target/ directory
```
Comment (duncanpMS) [inline Tasks/Maven/maventask.ts]: As discussed offline, this needs a more explanatory name e.g. _findCandidateModules_.

The comment should also call out that we have no way of knowing which POMs are actually being built; we are assuming that the user/POM hasn't changed the default _target_ build directory (which is possible), and identifying possible modules location by the presence of a pom.xml file and a /target/ directory.

Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Accepted.

File: Tasks/Maven/maventask.ts (hunk 5 above)
Comment (duncanpMS) [inline Tasks/Maven/maventask.ts]: Document the decision around grouping files for upload to provide a better user experience in the artefact explorer.

File: Tasks/Maven/maventask.ts (hunk 21 above)
Comment (duncanpMS) [inline Tasks/Maven/maventask.ts]: A simple example of an example summary as a comment in the code would help.

File: Tasks/Maven/moduleanalysis.ts (hunk 8 above)
Comment (duncanpMS) [inline Tasks/Maven/moduleanalysis.ts]: "// a dictionary of _tool name_ -> AnalysisResult" would be better than "// To be used as a dictionary"

File: Tasks/Maven/pmdformaven.ts (hunk 9 above)
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: As discussed offline, there doesn't seem to be any advantage to adding _-DlinkXRef=false_ unless the omitting it results in a warning/error.

File: Tasks/Maven/maventask.ts (hunk 21 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: This results in a "[WARNING] Unable to locate Source XRef to link to - DISABLED", but omitting it does allow the user to manually enable the "jxr:jxr" step and add the cross-referencing data if they want to.

File: Tasks/Maven/pmdformaven.ts [hunk 23]
```diff
@@ -10,25 +10,27 @@ import trm = require('vsts-task-lib/toolrunner');
...
 // Adds PMD goals, if selected by the user
 export function applyPmdArgs(mvnRun: trm.ToolRunner):void {
-    mvnRun.arg(['jxr:jxr', 'pmd:pmd']); // JXR source code cross-referencing is a required pre-goal for PMD
+    mvnRun.arg(['pmd:pmd', '-DlinkXRef=false']); // Turn off cross-referencing to reduce Maven error output
 }
 
 // Extract analysis results from PMD output file.
-// Takes the current working directory and returns an AnalysisResult data class.
+// Takes the working directory (should contain pom.xml and target/) and returns an AnalysisResult data class.
 // Task fails if the HTML or XML outputs were not found.
```
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: Line 20 - this comment is no longer accurate, is it?

Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 10 above)
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: Worth calling out the expected behaviour of the PMD plugin i.e. "We are expecting the PMD plugin to drop the analysis results files with well-known names in the default target directory."

File: Tasks/Maven/pmdformaven.ts [hunk 24]
```diff
@@ -10,25 +10,27 @@ import trm = require('vsts-task-lib/toolrunner');
...
+    result.toolName = toolName;
 
-    var pmdXmlFilePath = path.join(sourcesDirectory, '/target/pmd.xml');
+    var pmdXmlFilePath = path.join(rootDir, 'target', 'pmd.xml');
     result = processPmdXml(result, pmdXmlFilePath);
 
     // if there are no violations, there will be no HTML report
     if (result.totalViolations > 0) {
-        var pmdHtmlFilePath = path.join(sourcesDirectory, '/target/site/pmd.html');
+        var pmdHtmlFilePath = path.join(rootDir, 'target', 'site', 'pmd.html');
```
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: This appears to be the only point that the .xml and html files are treated differently depending on which file it is i.e. the _AnalysisResult_ should just have a list of "FilesToUpload" which both can be added to.

File: Tasks/Maven/maventask.ts (hunk 5 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Accepted.

File: Tasks/Maven/moduleanalysis.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Maven/moduleanalysis.ts]: Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 10 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 13 above)
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: As discussed, the handling should be consistent with the error case in lines 63-65 below - either ignore or throw in both cases. Should probably throw - we're not expecting it to happen, and if we don't throw we won't know that something is wrong.

File: Tests/L0/Maven/_suite.ts (hunk 17 above)
Comment (duncanpMS) [inline Tests/L0/Maven/_suite.ts]: Are there any tests that the build summary contains the expected data?

File: Tasks/Maven/pmdformaven.ts (hunk 9 above)
Comment (duncanpMS) [inline Tasks/Maven/pmdformaven.ts]: Hhhmm. That's a bit trickier. I'd suggest omitting it for now and Jean-Marc can decide whether the user experience is acceptable.

File: Tasks/Maven/pmdformaven.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Accepted.

File: Tests/L0/Maven/_suite.ts (hunk 17 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: The content of the uploaded build summary file is not checked by the tests.

Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved, to be discussed later.

File: Tasks/Maven/pmdformaven.ts (hunk 24 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Accepted.

File: Tasks/Maven/pmdformaven.ts (hunk 23 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved.

File: Tasks/Maven/pmdformaven.ts (hunk 24 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved.

File: Tasks/Maven/maventask.ts (hunk 22 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Resolved.

File: Tasks/Maven/maventask.ts (hunk 21 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Resolved.

File: Tests/L0/Maven/_suite.ts (hunk 17 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Resolved.

File: Tasks/Maven/maventask.ts (hunk 5 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Resolved.

File: Tasks/Maven/moduleanalysis.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Maven/moduleanalysis.ts]: Resolved.

File: Tasks/Maven/pmdformaven.ts (hunk 10 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved.

File: Tasks/Maven/pmdformaven.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved.

File: Tasks/Maven/pmdformaven.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: From an offline discussion with JM - we'll disable it for now, without adding the linkXRef flag.

Accepted.

Comment (HSAR) [inline Tasks/Maven/pmdformaven.ts]: Resolved, complete.

File: Tasks/Maven/maventask.ts [hunk 25]
```diff
@@ -273,23 +273,27 @@ function processMavenOutput(data) {
...
 }
 
-// Identifies all maven modules below the root by the presence of a pom.xml file and a /target/ directory
-function findModules(folder:string):ma.ModuleAnalysis[] {
+// Identifies maven modules below the root by the presence of a pom.xml file and a /target/ directory,
+// which is the conventional format of a Maven module.
+// There is a possibility of both false positives if the above two factors are identified in a directory
+// that is not an actual Maven module, or if the module is not currently being built.
+// The possibility of false positives should be taken into account when this method is called.
+function findCandidateModules(directory:string):ma.ModuleAnalysis[] {
```
Comment (bgavrilMS) [inline Tasks/Maven/maventask.ts]: I remember discussing that all the PMD related logic should be in a separate file.

File: Tasks/Maven/maventask.ts [hunk 26]
```diff
@@ -416,55 +410,75 @@ function uploadCodeAnalysisResults():void {
...
-            var buildSummaryFilePath:string = path.join(stagingDir, 'CodeAnalysisBuildSummary.md');
-            fs.writeFileSync(buildSummaryFilePath, buildSummaryContents);
-            tl.debug('Uploading build summary from ' + buildSummaryFilePath);
-
-            tl.command('task.addattachment', {
-                'type': 'Distributedtask.Core.Summary',
-                'name': "Code Analysis Report"
-            }, buildSummaryFilePath);
+            } catch (error) {
+                tl.error(error.message);
```
Comment (bgavrilMS) [inline Tasks/Maven/maventask.ts]: Why did you add the try catch?

**Comment by bgavrilMS** [comment]
LGTM

File: Tasks/Maven/maventask.ts (hunk 25 above)
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: We did, but it should go into a different pull request since this one is already fairly complex.
